src/index/Vocabulary.h

```diff
@@ -70,9 +72,27 @@ class Vocabulary {
   //! Append a word to the vocabulary. Wraps the std::vector method.
   void push_back(const string& word) { _words.push_back(word); }
 
-  //! Get the word with the given id (as const, not as lvalue).
-  const string& operator[](Id id) const {
-    return _words[static_cast<size_t>(id)];
+  //! Get the word with the given id or an empty optional if the
+  //! word is not in the vocabulary.
+  //! Note: So as to keep things as references this does *NOT* check the
+  //! externalized part of the vocabulary
+  const std::optional<std::reference_wrapper<const string>> operator[](
+      Id id) const {
+    if (id < _words.size()) {
+      return _words[static_cast<size_t>(id)];
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  const std::optional<string> idToOptionalString(Id id) const {
+    std::optional<string> res = (*this)[id];
+    if (!res && id != ID_NO_VALUE) {
+      id -= _words.size();
+      AD_CHECK(id < _externalLiterals.size());
+      res = _externalLiterals[id];
+    }
+    return res;
```

Comment:
I know how this works, and why you did this, but having ID_NO_VALUE in the Vocabulary class is unexpected to me:
This is a special property of the optional operator as I expected it and so I think the following should rather be in the class handling the optional: 
```
if (id == ID_NO_VALUE) {
  return std::nullopt;
} else {
  return _vocab.idToString(id) // or operator[] depending on what can be used.
}
```
I'm sorry I didn't see this the first time.

Reply:
I understand why you feel that it is unexpected. The problem is that `idToString(id)` is called pretty much throughout the codebase whenever we go from id to a string and every time we do this it might be a non existant optional value.

Reply:
Okay, this is indeed a good argument.